src/components/chatbot/ChatBox.tsx

```diff
@@ -0,0 +1,64 @@
+'use client'
+
+import React, { useEffect, useRef, useState } from 'react'
+import ChatBoxForm, { formSchema } from './ChatBoxForm';
+import ChatMessage from './ChatMessage';
+import { useChat } from '@/hooks/useChat';
+import { useSendMessage } from '@/hooks/useBot';
+import { ByUser, Message } from '@/types/messages/message';
+import { getMessagesFromChat } from '@/lib/chat';
+import { getMessageWebsocketUrl } from '@/constants/global';
+import { z } from 'zod';
+import { useWebsocket } from '@/hooks/useWebsocket';
+
+type P = {
+  // ID of the chat to retrieve history of messages from. Not the TG Chat ID, but the actual chat Id stored in database
+  chatId: number;
+}
+
+export default function ChatBox({chatId}: P) {
+  const {data: chat, isPending} = useChat(chatId);
+  const sendMessageMutation = useSendMessage(chatId)
+
+  const [messages, setMessages] = useState<Message[]>([])
+  const bottomRef = useRef<any>(null);
+  
+  useWebsocket(getMessageWebsocketUrl(), setMessages)
+
+  useEffect(() => {
+    if (chat) {
+      setMessages(getMessagesFromChat(chat));
+    }
+  }, [chat])
+
+  // Auto-scrol to bottom of chat upon new messages
+  useEffect(() => {
+    bottomRef.current?.scrollIntoView({ behavior: "smooth" });
+  }, [messages])
+
+  const onSubmit = async (values: z.infer<typeof formSchema>) => {
+    await sendMessageMutation.mutate(values.text);
+    setMessages([...messages, {
+      messageBody: values.text,
+      by: ByUser.STAFF,
+      telegramMessageId: 0,
+      timestamp: new Date(),
+    }])
```

Comment:
It seems like you are updating the UI regardless of the outcome (success/error) of the mutation.

I was thinking you can pass `onSuccess` and `onError` callbacks to useSendMessage hook (which passes the callback to useMutation). `onError` can be some toast notification

Also if you want to take it a step further, it would be a good use case for [useOptimistic hooks](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#optimistic-updates)

Reply:
My bad yes. It slipped my mind to worry about this desync issue.

Reply:
@wxiaoyun  Actually my idea was to just remove the setMessages call completely, because the react-query would automatically pull the latest messages. So we wouldn't need to set the messages state manually

However, potential downside is that we would have to wait for the server to finish processing the message before it shows on the UI side. What do you think?

Reply:
I agree thats the superior approach with cache invalidation and react query auto refetching :+1: